### PR TITLE
ExPlat: Self-Contained Client API

### PR DESCRIPTION
**Pull request by jessie-ross**
**This PR implements and unit tests the client API:**
- Includes the SSR safety mechanism and related tests.

Part of a series rebuilding the reference WIP PR #48475

#### Testing instructions

* Run `yarn test` within `/packages/explat-client`

### CODE REVIEW

File: packages/explat-client/src/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/create-explat-client.ts]: All this nesting of try-catch blocks isn't great... It seems like the main concern is that `logError()` would throw an error. To work around it, how about we add a local function like this:

```typescript
const safeLogError = ( message ) => {
  try {
    config.logError( { message, experimentName } );
  } catch ( e )  {
    if ( config.isDevelopmentMode ) {
      throw e;
    }
  }
}
```

Then, instead of the nested try-catch blocks we can do:

```typescript
  ...
  return fetchedExperimentAssignment;
} catch ( e ) {
  safeLogError( e.message );
  if ( config.isDevelopmentMode ) {
     throw e;
  }
}
// Fetching failed and we're not in development mode.
try {
  // ... fallback 1 code (lines 134-148)
} catch ( e ) {
  // ... fallback 2 code, with safeLogError()
}
```

Also, do we even need the second try-catch? Lines 134-148 shouldn't throw exceptions, should they?

File: packages/explat-client/src/create-ssr-safe-mock-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/create-ssr-safe-mock-explat-client.ts]: How about we call this `createSsrSafeDummyExPlatClient()`? There is no mocking in the test-mock sense.

Also, it'd be good to document this, and I think it can live in the `create-explat-client.ts` file.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: These lines should be removed now that we're using the full config.

```suggestion
```

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: It'd be more self-documenting to add a couple of helpers `setBrowserContext()` and `setSsrContext()` that'd change the `window` value instead of the three lines to override the window in multiple places. They can also be combined into a single helper like:

```typescript
function setBrowserContext( browserContext: boolean ) {
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	global.window = browserContext ? {} : undefined;
}
```

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: I think it'd be a good idea to expect something instead of suppressing the `expect-expect` warning. For example, we can check that the return value of each `createExPlatClient()` call isn't null.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: What's the reasoning behind sharing a global mock between tests rather than creating a unique mock for each test? It feels a bit brittle, and I'm also unsure why we're naming the mocked functions as separate consts.

Instead, we can do something like this: 
```typescript
function createConfigMock( overrides: Partial<Config> ): Config {
	return {
		logError: jest.fn(),
		fetchExperimentAssignment: jest.fn(),
		getAnonId: jest.fn(),
		isDevelopmentMethod: false,
		...overrides
	}
}
```

With this, we can do stuff like `configMock = createConfigMock()` and  `configMock.getAnonId.mockImplementationOnce(...)`, or pass in the specific mock overrides to the creation function.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: As noted on another PR, these `delayedValue()` calls are hard to reason about and should at least be documented, though constants would be better than magic numbers.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: Why do we care what calls the client made as long as it returns what we expect based on the mocks? These sort of checks may make the implementation too rigid.

It seems like the only mock that we should generally check is `logError()`, since it's a side-effect that we normally care about.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: This is a good example of a case where we really shouldn't care about the mocks called. Eliminating such cases would make this big test file much easier to digest and maintain.

I'll hold off on reading the rest until we've discussed the points made so far.

File: packages/explat-client/src/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/create-explat-client.ts]: My idea here is to make this failsafe, and the second try-catch makes that simpler: if we have any code here we should assume it can go wrong and have a backup plan. If we push any failures into the backup plan we then only need to ensure the backup plan itself can't go wrong rather than the rest of the code.

In this case the backup plan is the final `createFallbackExperimentAssignment`.


> Lines 134-148 shouldn't throw exceptions, should they?

They shouldn't but now we are covered in the case they do.

Also it is not just `logError` but also `storeExperimentAssignment` - the internals can also throw.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: Ah I did this to make `mockFetchExperimentAssignmentToMatchExperimentAssignment` easier to use but fair enough, not too hard to make it work.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: To me this is integration testing, ensuring the boundary of the client stays constant.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: If the size of the file is a concern I can split it by the top level describes.

File: packages/explat-client/src/create-explat-client.ts
Comment (yanirs) [reply packages/explat-client/src/create-explat-client.ts]: OK, it seems a bit too defensive to me, but I still think that flattening the try-catches is worthwhile. Having three nested levels of try-catch blocks with error names like `e`, `e2`, and `e3` is harder to follow than the flow I suggested, which makes it clear that we only get into fallbacks outside dev mode.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [reply packages/explat-client/src/test/create-explat-client.ts]: Yeah, I was thinking we could also make the overrides more sophisticated so that passing an `ExperimentAssignment` object as the value of `fetchExperimentAssignment` would also create the mock, but I'm not sure it'd be that clean without playing with it.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [reply packages/explat-client/src/test/create-explat-client.ts]: I'm not sure that this counts as the boundary of the client, as the exact internal calls aren't a part of the contract. In many cases, the fact that the calls happened is implied by the observed outputs.

The main output that we don't observe is that of `logError()`, which is worth checking when we expect errors.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [reply packages/explat-client/src/test/create-explat-client.ts]: It's just a lot to review &ndash; I don't think we need to split it to files, just ensure that the tests only test what they need to test. Having less code to read/maintain is a nice side effect. :slightly_smiling_face: 

My point still stands: In this case, we don't care about the mocks since an error was thrown, so I think we should remove the mock check. This applies to other cases where the behaviour is enough to tell us that the mocks were used as expected.

We can get a second opinion on this from people who work in testing, though.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: So I have added `createConfigMock` so far but now I am thinking that this doesn't add much - mocks are reset anyway in-between tests, and making this work well with typescript is a bit of a pain as it needs to fit the two different uses - as a Config and as a mocked Config (e.g. to access `.mockImplement()`).

File: packages/explat-client/src/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/create-explat-client.ts]: This should be removed as it's unreachable (we're guaranteed not to be in development mode at this point).

File: packages/explat-client/src/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/create-explat-client.ts]: `retreive` -> `retrieve`

File: packages/explat-client/src/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/create-explat-client.ts]: Is the `isSecondaryError` value going to be read by anything? If not, let's just change the message, e.g., the message here can be `'Fallback Error: ' + e.message`.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts:316]: If I understand the test case description correctly, there should be a secondary error logged, right?

File: packages/explat-client/src/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/create-explat-client.ts]: As [suggested previously](https://github.com/Automattic/wp-calypso/pull/49832#discussion_r572594842), I think that adding a thin wrapper around `logError` (aka `safeLogError`) would make this much more readable.

Alternatively, giving the exceptions more informative names would probably help. Something like:

* First `e`: `initialFetchError`
* Second `e`: `fallbackError`
* First and second `e2`: `ignoredLoggingError`

File: packages/explat-client/src/test/index.ts
Comment (yanirs) [inline packages/explat-client/src/test/index.ts]: Should use the `setBrowserContext()` and `setSsrContext()` helpers in this file too.

File: packages/explat-client/src/test/index.ts
Comment (yanirs) [inline packages/explat-client/src/test/index.ts:26]: Wouldn't this work without the `toString()` calls?

File: packages/explat-client/src/test/create-ssr-safe-mock-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-ssr-safe-mock-explat-client.ts]: These lines are unnecessary now.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: Continuing the discussion from https://github.com/Automattic/wp-calypso/pull/49832#discussion_r572610336, this sharing of globals across test cases is still problematic, and the current setup is confusing. When looking at an isolated test case, it's not immediately obvious that a name like `mockedGetAnonId` is related to the mock config we just created, and declaring `DEVELOPMENT_MODE = true` at the top of the file sounds like it's setting development mode to true.

Can you please explain further why the suggestion I made won't work? Copying here for context:

```typescript
function createConfigMock( overrides: Partial<Config> ): Config {
	return {
		logError: jest.fn(),
		fetchExperimentAssignment: jest.fn(),
		getAnonId: jest.fn(),
		isDevelopmentMethod: false,
		...overrides
	}
}
```

With this, we can do stuff like `configMock = createConfigMock()` and  `configMock.getAnonId.mockImplementationOnce(...)`, or pass in the specific mock overrides to the creation function.

If TypeScript is unhappy with some of the mocking, we can suppress specific warnings for the file / function.

I can also play with it if you prefer.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: Shouldn't we check all the dependencies to match the test description?

We can also drop this test case completely, as I don't think it's very interesting &ndash; we get plenty of coverage from the other tests.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: For this test to match our expectations, we should set up the mock so that it only returns the experiment assignment if the anon ID is `the-anon-id-123`.

If we make the mocking changes suggested above, it'd be something like `mockFetchExperimentAssignmentToMatchExperimentAssignment( { experimentAssignment: validExperimentAssignment, expectedAnonId: 'the-anon-id-123' } )`

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: This mock function doesn't match the signature of the function we're mocking. We should accept the `experimentName` and `anonId` parameters and check their value. The formatting is a mess, but I think that something like this would work:

```typescript
function mockFetchExperimentAssignmentToMatchExperimentAssignment(
	{ experimentAssignment, expectedAnonId } :
	{ experimentAssignment: ExperimentAssignment; expectedAnonId: string | null; }
) {
	spiedMonotonicNow.mockImplementationOnce( () => experimentAssignment.retrievedTimestamp );
	mockedFetchExperimentAssignment.mockImplementationOnce( ( experimentName, anonId ) => {
		expect( experimentName ).toBe( experimentAssignment.experimentName );
		expect( anonId ).toBe( expectedAnonId );
		...
```

See also below for how it's useful in the context of specific tests.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: As ntoed on #49829, I think there's no harm in returning the invalid name here. It's less surprising than `fallback_experiment_assignment`.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts:228]: Not critical, but this file may be easier to follow if we split the development mode tests to nested `describe`s. That is:

```typescript
describe( 'ExPlatClient.loadExperimentAssignment single-use', () => {
	describe( 'in production mode, () => {
		// ... production mode tests ...
	} );

	describe( 'in development mode, () => {
		// ... development mode tests ...
	} );
} );
```

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: This is almost identical to the previous test. Is the behaviour supposed to be the same? It'd be good to reduce duplication here to make the intention clearer.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [inline packages/explat-client/src/test/create-explat-client.ts]: It'd be nice to add an `awaitException()` helper that'd return the error. Then we can do `const errorA = awaitException( experimentAssignmentA )`. The helper can also fail if the exception isn't thrown.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [reply packages/explat-client/src/test/create-explat-client.ts]: Ah! On further reflection, I think that a part of my confusion here is that technically, we're not mocking the config. We are using Jest's mocking helpers to create a valid config, which is passed to the `createExPlatClient()` function. In general, clients should be free to put whatever they want in the config, as long as it adheres to the interface.

This is just an aside, though. I still think that reducing the reliance on globals would make the tests clearer.

File: packages/explat-client/src/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/create-explat-client.ts]: Yeah it ends up LogStashed, I think it is useful to diagnose errors in the wild.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: See the comment there - it is not a valid ExperimentAssignment and would fail various run-time checks.

File: packages/explat-client/src/create-explat-client.ts
Comment (yanirs) [reply packages/explat-client/src/create-explat-client.ts]: Maybe we can make it more descriptive then? It looks odd that we're passing `'true'` as a string, but I see that's needed from `logError`'s signature. We can add something like `context: 'failed fallback attempt'` and `context: 'failed fetch attempt'` for the initial try.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: I have replied there and have also changed the name to indicate that it is an invalid name.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts:228]: I was thinking about this but I like that the development and production tests are next to each other - it is easier to compare behaviour, makes refactoring a bit more obvious, and makes an omissions of a corresponding developer test clear.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: I have changed it closer to what you suggested and it didn't end up being so bad.

File: packages/explat-client/src/test/index.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/index.ts:26]: No it doesn't seem to, and this seemed to work well enough for our purposes although I can understand it isn't ideal.

Comment (jessie-ross) [reply packages/explat-client/src/test/index.ts:26]: Screen Shot 2021-02-18 at 4 00 02 pm

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: I have just ended up checking the `mockedConfig.requestExperimentAssignment.mock.calls`.

File: packages/explat-client/src/test/index.ts
Comment (yanirs) [reply packages/explat-client/src/test/index.ts:26]: Ah, I guess it may be because of the way imports work... Anyway, I agreed that using `toString()` is good enough.

File: packages/explat-client/src/test/index.ts
Comment (yanirs) [inline packages/explat-client/src/test/index.ts]: It'd be good to put these functions in the common utility file rather than duplicate them.

File: packages/explat-client/src/test/create-explat-client.ts
Comment (yanirs) [reply packages/explat-client/src/test/create-explat-client.ts]: Why? How is it better than using a more realistic mock?

**Review by yanirs — APPROVED** [review]
I still think it's worth addressing https://github.com/Automattic/wp-calypso/pull/49832#discussion_r578219835 and reducing duplication a bit, but should be good to go along with the rest of the stack (and code style fixes). :ship:

File: packages/explat-client/src/test/create-explat-client.ts
Comment (jessie-ross) [reply packages/explat-client/src/test/create-explat-client.ts]: Screen Shot 2021-02-18 at 4 58 17 pm

It gets a bit complicated as jest works by throwing errors and the error is occurring within a config function. I am also sometimes returning a different experimentAssignment to what was requested to check other things.

I could still add more checks on the requestExperimentAssignment calls, maybe snapshotting them?

Comment (yanirs) [reply packages/explat-client/src/test/create-explat-client.ts]: Ah, never mind then. I think what we have is good enough.

**Pull request merged**
